## NumPy MLP — data preparation and baseline implementation
In this notebook we include data preparation steps, a simple NumPy MLP implementation, training routines, and a comparison with sklearn.MLPClassifier.

### Prerequisites
First you need to unzip data/*.zip

In [47]:
import pandas as pd

In [48]:
train = pd.read_csv('data/train.csv')
test = pd.read_csv('data/test.csv')


In [49]:
print(train.shape)
print(test.shape)


(42000, 785)
(28000, 784)


In [50]:
print(train.sample(1))
print(test.sample(1))

       label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
38448      1       0       0       0       0       0       0       0       0   

       pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  \
38448       0  ...         0         0         0         0         0   

       pixel779  pixel780  pixel781  pixel782  pixel783  
38448         0         0         0         0         0  

[1 rows x 785 columns]
       pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
20665       0       0       0       0       0       0       0       0       0   

       pixel9  ...  pixel774  pixel775  pixel776  pixel777  pixel778  \
20665       0  ...         0         0         0         0         0   

       pixel779  pixel780  pixel781  pixel782  pixel783  
20665         0         0         0         0         0  

[1 rows x 784 columns]


In [51]:
print(sorted(train['label'].unique()))
print(sorted(train['label'].value_counts()))

[np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9)]
[3795, 4063, 4072, 4132, 4137, 4177, 4188, 4351, 4401, 4684]


In [52]:
print(train['pixel394'].unique())

[  0  29 223  18 192  26  72  84 115  37  86  54 254 162  90 124 228 128
 247  13  22]


### Conclusion
- We have 784 pixels 28x28
- Target is first column (label)  
- Target is digit from 0 to 9 - 10 unique values
- All the digits appears similarly the most difference is between 0 and 9 (3795 x 4684) 
- Values in pixelX column is form 0 to 255 meaning intensivity 